### PR TITLE
PubSubManager for Event Logging

### PR DESCRIPTION
**Pull request by vudung45**
Usage example
```javascript
// define constants, these values are singletons or there are only very few of them defined in the entire codebase
const GARoundtripHandler = (trackerInfo: { eventName: string, timeFired: number, aux: any}) => {
   // handle send event logger to GA
})
const pubSubManager = new PubSubManager();
pubSubManager.register(socket, "message");
pubSubManager.register(socket, "join_room");
pubSubManager.register(socket, "global_message");

// add triggers to be called on event.publish()
pubSubManager.addTrigger("message", GAImpressionLogHandler /*GA call for impression logging*/); 

....
//  some other parts of the codebase
pubSubManager.publish("report_user", {"username" : "davidvu" }); // no roundtrip tracker
// or
pubSubManager.publish("message", {"messasge" : "hi there" }, GARoundtripHandler); // GARoundtripHandler will be executed when subscribe("message", ...) receives a callback

...
// subscribe to event
pubSubManager.subscribe("message", (payload) => {
   // add message to chatbox
}, track = true); // with track = true, GARoundtripHandler() is called, else it would just behave as a normal subscriber


```

### CODE REVIEW

**Comment by namdaoduy** [comment]
@vudung45 As I can see, each type of event will need a different eventTracker?
Because event data is very depends on type of event. Combine them in one tracker will make the code quite hard to read.

**Comment by vudung45** [comment]
@namdaoduy Nam
There should be only 1 GAHandler so make them as generic as possible
e.g for GA roundtrip Logger
```
const GAHandler = (eventTracker: { eventName: string, timeFired: number, aux: any}) => {
   ReactGA.event({
          category: `${getEventCategory(eventName)}_roundtrip`,
          action: getEventType(eventName),
          label: Date.now() - eventName,
        });
})
```
GAHandler is fired inside subscriber callback.

File: frontend/src/Utils/EventTracker.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        triggersOnPublish: [],
+      };
+    }
+    if (triggerOnPublish)
+      this.eventStore[eventName].triggersOnPublish.push(triggerOnPublish);
+  }
+
+  subscribe(
+    eventName: string,
+    handler: (payload: Payload) => any,
```
Comment (namdaoduy) [inline frontend/src/Utils/EventTracker.ts:83]: handler can be undefined also. Some events also need to be emit.

File: frontend/src/Utils/EventTracker.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      pubSub.on?.(eventName, handler);
+      return;
+    }
+
+    let handlerWrapper = (payload: TrackablePayload) => {
+      let eventTrack = this.findEventSourceTrack(eventName, payload.trackId);
+      if (eventTrack) {
+        eventTrack.roundtripHandler(eventTrack.trackInfo);
+      }
+      return handler(payload);
```
Comment (namdaoduy) [inline frontend/src/Utils/EventTracker.ts:101]: Check if handler is undefined

**Review by namdaoduy — CHANGES_REQUESTED** [review]

File: frontend/src/Utils/EventTracker.ts (hunk 1 above)
Comment (vudung45) [reply frontend/src/Utils/EventTracker.ts:83]: subscribe is equivalent to socket.on()
User only calls it when they want to handle callback

File: frontend/src/Utils/EventTracker.ts (hunk 2 above)
Comment (vudung45) [reply frontend/src/Utils/EventTracker.ts:101]: ^

File: frontend/src/Utils/EventTracker.ts (hunk 1 above)
Comment (namdaoduy) [reply frontend/src/Utils/EventTracker.ts:83]: This usecase:
- User only want to emit an event, no handler needed for incoming event
- User want to track roundtrip on that event
=> Additional handler on `on` is unnecessary. I have to put empty funtion on that handler.
@vudung45

Comment (namdaoduy) [reply frontend/src/Utils/EventTracker.ts]: @vudung45 Sorry, i confused between `register` and `subscribe`. LGTM now!

**Review by namdaoduy — APPROVED** [review]

File: frontend/src/Utils/EventTracker.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      if (eventTrack) {
+        eventTrack.roundtripHandler(eventTrack.trackInfo);
+      }
+      return callback(payload);
+    };
+    pubSub.on?.(eventName, callbackWrapper);
+  }
+
+  /**
+   * Publish a new message to a channel
```
Comment (vudung45) [inline frontend/src/Utils/EventTracker.ts:120]:
```suggestion
   * Publish a new message to channel
```

File: frontend/src/Components/Chatbox.tsx [hunk 4]
```diff
@@ -254,17 +249,17 @@ export default function Chatbox({
...
+        socketTracker.unsubscribe("delete_message");
+        socketTracker.unsubscribe("report_applied");
+        socketTracker.unsubscribe("ban_applied");
+        socketTracker.unsubscribe("unban_applied");
+        socketTracker.unsubscribe("join_room_resp");
+        socketTracker.unsubscribe("report_user_resp");
       }
     };
-  }, [socket, userData.username]);
+  }, [socketTracker, userData.username]);
```
Comment (vudung45) [inline frontend/src/Components/Chatbox.tsx:262]: Let's only resubscribe when socketTracker changes! 

```suggestion
  }, [socketTracker]);
```

File: backend/userstate.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (manhhung741) [inline backend/userstate.json:1]: cai file nay tuong minh remove roi

Comment (manhhung741) [reply backend/userstate.json:1]: nho' remove file nay nha.

**Pull request merged**